src/coreclr/nativeaot/Runtime/GCMemoryHelpers.cpp

```diff
@@ -44,17 +44,30 @@ FCIMPLEND
 
 FCIMPL3(void, RhBulkMoveWithWriteBarrier, uint8_t* pDest, uint8_t* pSrc, size_t cbDest)
 {
-    // It is possible that the bulk write is publishing object references accessible so far only
-    // by the current thread to shared memory.
-    // The memory model requires that writes performed by current thread are observable no later
-    // than the writes that will actually publish the references.
-    GCHeapMemoryBarrier();
+    if (cbDest == 0)
```

Comment:
Did you run into any case where we are actually passing in 0 here?

Reply:
Yes, CI was all red because I deleted the previous "size == 0" check in the card table updater (`InlinedBulkWriteBarrier`)
```c
(cbMemSize < sizeof(uintptr_t)))
```
I'll check who exactly calls it with zero length

Reply:
@jkotas it looks like there is no protection from zero size in NAOT, e.g. in CoreCLR we have [this](https://github.com/dotnet/runtime/blob/dc982632aa2b15123ba5d7ff76589cd6c8e660ba/src/coreclr/vm/comutilnative.cpp#L528) (we even check whether pointers are the same), while in NAOT we just go straight to the copy algorithm. One example that fires here is `Array.Copy` for empty array

![image](https://github.com/dotnet/runtime/assets/523221/9c4c86d1-ddc6-4766-b1ae-f7583fb285d6)

Reply:
Ok, sounds good.